store/src/state.rs

```diff
@@ -102,14 +180,127 @@ impl State {
+        &self,
+        new_block: NewBlock,
+        nullifiers: &[RpoDigest],
+        accounts: &[(AccountId, Digest)],
+        notes: &[Note],
+    ) -> Result<Option<block_header::BlockHeader>, anyhow::Error> {
+        let _ = self.writer.try_lock().map_err(|_| StateError::ConcurrentWrite)?;
+
+        // ensures the right block header is being processed
+        let prev_block_msg = self
+            .db
+            .select_block_header_by_block_num(None)
+            .await?
+            .ok_or(StateError::DbBlockHeaderEmpty)?;
+        let block_num = (prev_block_msg.block_num as u32) + 1;
+        let prev_block: BlockHeader = prev_block_msg.try_into()?;
+        let prev_hash = prev_block.hash();
+        if new_block.prev_hash != prev_hash {
+            return Err(StateError::NewBlockInvalidPrevHash.into());
+        }
+
+        // scope to read in-memory data, validate the request, and compute intermediary values
+        let (account_tree, chain_mmr, nullifier_tree, block_header) = {
+            let inner = self.inner.read().await;
+
+            // nullifiers can be produced only once
+            let duplicate_nullifiers: Vec<_> = nullifiers
+                .iter()
+                .filter(|&&n| inner.nullifier_tree.get_value(n) != EMPTY_WORD)
+                .cloned()
+                .collect();
+            if !duplicate_nullifiers.is_empty() {
+                return Err(StateError::DuplicatedNullifiers(duplicate_nullifiers).into());
+            }
+
+            // update the in-memory datastructures and compute the new block header. Important, the
+            // structures are not yet commited
+            let mut chain_mmr = inner.chain_mmr.clone();
+            chain_mmr.add(prev_hash);
+            let peaks = chain_mmr.peaks(chain_mmr.forest())?;
+
+            let mut nullifier_tree = inner.nullifier_tree.clone();
+            let nullifier_data = block_to_nullifier_data(block_num);
+            for nullifier in nullifiers {
+                nullifier_tree.insert(*nullifier, nullifier_data);
+            }
+
+            let mut account_tree = inner.account_tree.clone();
+            for (account_id, account_hash) in accounts {
+                account_tree.update_leaf(*account_id, account_hash.try_into()?)?;
+            }
+
+            let note_tree = build_notes_tree(notes)?;
+
+            let block_header = block_header::BlockHeader {
+                prev_hash: Some(prev_hash.into()),
+                block_num,
+                chain_root: Some(peaks.hash_peaks().into()),
+                account_root: Some(account_tree.root().into()),
+                nullifier_root: Some(nullifier_tree.root().into()),
+                note_root: Some(note_tree.root().into()),
+                batch_root: Some(new_block.batch_root.into()),
+                proof_hash: Some(new_block.proof_hash.into()),
+                version: new_block.version,
+                timestamp: new_block.timestamp,
+            };
```

Comment:
Could we use `BlockHeader` from `miden-objects` here?

Reply:
this code no longer exists because of the request in this comment: https://github.com/0xPolygonMiden/miden-node/pull/70#discussion_r1405341692